illumos-utils/src/contract.rs

```diff
@@ -0,0 +1,387 @@
+#[cfg(target_os = "illumos")]
+use libc::ctid_t;
+use slog::{Logger, debug, error};
+use std::ffi::CString;
+use std::ffi::c_char;
+use std::ffi::c_int;
+use std::ffi::c_uint;
+use std::ffi::c_void;
+use std::thread;
+use std::time::Duration;
+
+#[allow(non_camel_case_types)]
+type ct_evthdl_t = *mut c_void;
+#[allow(non_camel_case_types)]
+type ct_evid_t = u64;
+
+#[link(name = "contract")]
+unsafe extern "C" {
+    fn ct_tmpl_set_critical(fd: c_int, events: c_uint) -> c_int;
+    fn ct_tmpl_set_informative(fd: c_int, events: c_uint) -> c_int;
+    fn ct_pr_tmpl_set_fatal(fd: c_int, events: c_uint) -> c_int;
+    fn ct_pr_tmpl_set_param(fd: c_int, params: c_uint) -> c_int;
+    fn ct_dev_tmpl_set_minor(fd: c_int, minor: *const c_char) -> c_int;
+    fn ct_tmpl_activate(fd: c_int) -> c_int;
+    fn ct_tmpl_create(fd: c_int, ctid: *mut c_int) -> c_int;
+    fn ct_tmpl_clear(fd: c_int) -> c_int;
+    fn ct_ctl_abandon(fd: c_int) -> c_int;
+    fn ct_ctl_ack(fd: c_int, evid: ct_evid_t) -> c_int;
+    fn ct_event_read_critical(fd: c_int, ev: *mut ct_evthdl_t) -> c_int;
+    fn ct_event_get_type(ev: ct_evthdl_t) -> u32;
+    fn ct_event_get_ctid(ev: ct_evthdl_t) -> ctid_t;
+    fn ct_event_get_evid(ev: ct_evthdl_t) -> ct_evid_t;
+    fn ct_event_free(ev: ct_evthdl_t);
+}
+
+// Convert an error message into an ExecutionError::ContractFailure
+fn err(msg: impl ToString) -> crate::ExecutionError {
+    return crate::ExecutionError::ContractFailure {
+        msg: msg.to_string(),
+        err: std::io::Error::last_os_error(),
+    };
+}
+
+// Construct a path to a file in the contract filesystem
+fn path(typ: ContractType, id: Option<c_int>, file: &str) -> CString {
```

Comment:
Looking at the call sites of this function, I'm not sure it's really pulling its weight. In particular, if the calls to `libc::open()` were replaced with calls to `nix::fcntl::open` (as mentioned below) then the most interesting bit is the match on `typ` to set the prefix. But I'd argue that that's a good candidate for a method on `ContractType`:

```rust
impl ContractType {
    pub fn path_prefix(self) -> PathBuf {
        PathBuf::from(match self {
            Self::Process => "/system/contract/process",
            Self::Device => "/system/contract/device",
        })
    }
}
```

Reply:
Yep.  Better.  Thanks.